org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -0,0 +1,495 @@
+package org.mozilla.jss.ssl.javax;
+
+import java.io.*;
+import java.net.*;
+import java.nio.ByteBuffer;
+import java.nio.channels.*;
+import java.security.*;
+import java.util.*;
+
+import javax.net.ssl.*;
+
+/**
+ * SSL-enabled SocketChannel following the javax.net.ssl.SSLSocket interface.
+ *
+ * This class should never be constructed directly and instead only accessed
+ * once a Socket is wrapped in a JSSSocket.
```

Comment:
So we're not creating a `SocketChannel` here with `JSSSocketChannel`, just like we're not creating a `Socket` in the `JSSSocket` -- we're creating what _would_ be a `SSLSocketChannel` implementation. But such a thing doesn't (apparently) exist in the standard library. 

We're upgrading/"consuming" an existing `SocketChannel` and `Socket`, wrapping it in `JSSSocketChannel` and `JSSSocket`, and making the result do TLS over this wrapped `Socket`/`SocketChannel`. 

---

Part of the problem here is we need to blend two worlds: the blocking world of `SSLSocket` and the non-blocking world of `SSLEngine`. I chose to take the consumed `Socket` and convert the `InputStream`/`OutputStream`s into `Channel`s so that we can more easily utilize them with `SSLEngine`. Otherwise we'd have to put `byte[]` buffers directly inside `JSSSocket` and `.wrap` them to `ByteBuffer`s for use with `SSLEngine`.

We're not trying to directly mimic what the SunJSSE `SSLSocketImpl` does here either. It has very restrictive usages that the original `org.mozilla.jss.ssl.SSLSocket` doesn't have and we don't need either:

 1. If your socket isn't connected, SunJSSE's SocketFactory won't consume it via `createSocket(Socket, <host>, <port>, <auto close>)` -- it'll fail saying that `Socket` should've been connected first. We instead create an unconnected socket and leave it to the caller to call `connect` the new socket.
 2. If your socket is non-blocking, it'll fail to work. `startHandshake()` will fail with an invalid blocking mode exception. This is because they internally didn't implement any channel semantics.

Since we can use the JSSSocketChannel to provide both blocking and non-blocking semantics, I think we should. It lets people use a non-blocking channel without manually maintaining a SSLEngine instance. It just requires them to go through a SSLSocketFactory in order to do it, which I think is fine.

Reply:
@edewata  Do you think #2 makes sense? Should I also create a generic `SSLSocketChannel` that brings some of the methods from `SSLSocket` onto a `SSLSocketChannel`?

Reply:
These are tough questions! :)
I'd say go ahead and implement anything that we need to get it working,
and feel free to implement anything extra that will make JSS more usable
in general.